nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateJson.java

```diff
@@ -0,0 +1,215 @@
+
+    public static final PropertyDescriptor SCHEMA_TEXT = new PropertyDescriptor
+        .Builder().name("SCHEMA_TEXT")
+        .displayName("Schema Text")
+        .description("The text of a JSON schema")
+        .required(true)
+        .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .build();
+
+    public static final Relationship REL_VALID = new Relationship.Builder()
+        .name("valid")
+        .description("FlowFiles that are successfully validated against the schema are routed to this relationship")
+        .build();
+
+    public static final Relationship REL_INVALID = new Relationship.Builder()
+        .name("invalid")
+        .description("FlowFiles that are not valid according to the specified schema are routed to this relationship")
+        .build();
+
+    public static final Relationship REL_FAILURE = new Relationship.Builder()
+        .name("failure")
+        .description("FlowFiles that cannot be read as JSON are routed to this relationship")
+        .build();
+
+    private List<PropertyDescriptor> descriptors;
+
+    private Set<Relationship> relationships;
+
+    private List<AllowableValue> allowableValues;
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        final List<PropertyDescriptor> descriptors = new ArrayList<PropertyDescriptor>();
+        descriptors.add(SCHEMA_TEXT);
+        descriptors.add(SCHEMA_VERSION);
+        this.descriptors = Collections.unmodifiableList(descriptors);
+
+        final Set<Relationship> relationships = new HashSet<Relationship>();
+        relationships.add(REL_VALID);
+        relationships.add(REL_INVALID);
+        relationships.add(REL_FAILURE);
+        this.relationships = Collections.unmodifiableSet(relationships);
+
+        final List<AllowableValue> allowableValues = new ArrayList<AllowableValue>();
+        allowableValues.add(SCHEMA_VERSION_4);
+        allowableValues.add(SCHEMA_VERSION_6);
+        allowableValues.add(SCHEMA_VERSION_7);
+        allowableValues.add(SCHEMA_VERSION_V201909);
+        this.allowableValues = Collections.unmodifiableList(allowableValues);
+
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return this.relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if ( flowFile == null ) {
+            return;
+        }
+
+        final AtomicReference<Exception> exceptions = new AtomicReference<>();
+        final AtomicReference<Set<ValidationMessage>> validationErrors = new AtomicReference<Set<ValidationMessage>>(null);
+
+        session.read(flowFile, new InputStreamCallback()  {
```

Comment:
I'd recommend instead of using the InputStreamCallback pattern here using just `session.read()` that returns an InputStream. That way, the pattern looks something like:
```
try (InputStream in = session.read(flowFile)) {
   // processing logic
} catch (Exception e) {
  // Exception handling logic here
}
```
This pattern is nice because it avoids having to create a bunch of AtomicReference objects, which gets a little awkward to handle properly.

Reply:
Thanks, this is much easier - I didn't realise you could call session.read like this!